From 591198461af3bbf644b7b2a257a902f827c5fde6 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Thu, 7 Apr 2022 16:12:56 +0800
Subject: [PATCH] memcg: Fix inconsistent oom event behavior for OOM_MEMCG_KILL
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 591198461af3bbf644b7b2a257a902f827c5fde6
Modified-by-SEL: No


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I4X0YD?from=project-issue
CVE: NA

--------

Since memory.event is fully supported in cgroupv1, the problem of inconsistent
oom event behavior for OOM_MEMCG_KILL occurs again.
We fix the problem by add a new condition to support the event adding
continue. Therefore, there are two condition:
1) memcg is not root memcg;
2) the memcg is root memcg and the event is OOM_MEMCG_KILL of cgroupv1

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
Reviewed-by: Xiu Jianfeng <xiujianfeng@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 include/linux/memcontrol.h | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 23f4a3c8fef1..9d3a952024a6 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -1197,6 +1197,18 @@ static inline void count_memcg_event_mm(struct mm_struct *mm,
 	rcu_read_unlock();
 }
 
+static bool memcg_event_add(struct mem_cgroup *memcg,
+			    enum memcg_memory_event event)
+{
+	if (!mem_cgroup_is_root(memcg))
+		return true;
+
+	if (event == MEMCG_OOM_KILL && !cgroup_subsys_on_dfl(memory_cgrp_subsys))
+		return true;
+
+	return false;
+}
+
 static inline void memcg_memory_event(struct mem_cgroup *memcg,
 				      enum memcg_memory_event event)
 {
@@ -1217,7 +1229,7 @@ static inline void memcg_memory_event(struct mem_cgroup *memcg,
 		if (cgrp_dfl_root.flags & CGRP_ROOT_MEMORY_LOCAL_EVENTS)
 			break;
 	} while ((memcg = parent_mem_cgroup(memcg)) &&
-		 !mem_cgroup_is_root(memcg));
+		 memcg_event_add(memcg, event));
 }
 
 static inline void memcg_memory_event_mm(struct mm_struct *mm,
-- 
2.26.2

